src/vidpf.rs

```diff
@@ -0,0 +1,667 @@
+        cw_i: &CorrectionWord<R>,
+    ) -> R {
+        let CorrectionWord {
+            s_cw,
+            t_cw_l,
+            t_cw_r,
+            w_cw,
+        } = cw_i;
+
+        let seq_tilde = &self.prg(s_i);
+        let seq_cw = &Sequence(s_cw.clone(), *t_cw_l, s_cw.clone(), *t_cw_r);
+        let Sequence(sl, tl, sr, tr) = seq_tilde ^ (*t_i & seq_cw).borrow();
+
+        let s_tilde_i = Seed::conditional_select(&sl, &sr, alpha_i.0);
+        *t_i = ControlBit::conditional_select(&tl, &tr, alpha_i.0);
+
+        let w_i: R;
+        (*s_i, w_i) = self.convert(&s_tilde_i);
+
+        let mut y_i = R::conditional_select(&w_i, &(w_i.borrow() + w_cw.borrow()), t_i.0);
+        y_i.conditional_negate(Choice::from(b as u8));
+
+        y_i
+    }
+
+    /// `proof_next` generates a new [Proof] at the current level `(alpha,level)` from the
+    /// current state `(s_i,t_i,cs_i)` and current proof `(pi)`.
+    #[must_use]
+    pub fn proof_next(
+        &self,
+        pi: &Proof,
+        alpha: &[u8],
+        level: usize,
+        si: &Seed,
+        ti: ControlBit,
+        cs_i: &Proof,
+    ) -> Proof {
+        let pi_tilde = &Self::hash_one(alpha, level, si);
+        let h2_input = pi ^ (pi_tilde ^ (ti & cs_i).borrow()).borrow();
+        let out_pi = pi ^ Self::hash_two(&h2_input).borrow();
+        out_pi
+    }
+
+    /// Eval queries the evaluation tree with `alpha` input and produces a share of the output value.
+    ///
+    /// # Panics
+    /// Panics when the public key does not contain as many correction words as the length
+    /// of the `alpha` input in bits.
+    #[must_use]
+    pub fn eval(&self, alpha: &[u8], key: &Key, pk: &PublicKey<R>) -> Share<R> {
+        let n = 8 * alpha.len();
+        assert!(pk.cw.len() >= n, "bad public key size of cw field");
+        assert!(pk.cs.len() >= n, "bad public key size of cs field");
+
+        let mut s_i = Seed::from(key);
+        let mut t_i = ControlBit::from(key);
+
+        let mut y = R::new();
+        let mut pi = Proof::default();
+
+        for i in 0..n {
+            let alpha_i = ControlBit::from((alpha[i / 8] >> (i % 8)) & 0x1);
+            y = self.eval_next(key.id, alpha_i, &mut s_i, &mut t_i, &pk.cw[i]);
+            pi = self.proof_next(&pi, alpha, i, &s_i, t_i, &pk.cs[i]);
+        }
+
+        Share { y, pi }
+    }
+
+    /// verify checks whether the proofs are equal and that the shares add up to beta.
+    #[must_use]
+    pub fn verify(&self, a: &Share<R>, b: &Share<R>, beta: &R) -> bool {
+        a.y.add(&b.y) == *beta && a.pi.0 == b.pi.0
+    }
+
+    fn prg(&self, seed: &Seed) -> Sequence {
+        let dst = b"100";
+        let mut rng = XofFixedKeyAes128::seed_stream(&seed.0, dst, &self.bind_info);
+
+        let mut sl = Seed::default();
+        let mut sr = Seed::default();
+        rng.fill_bytes(&mut sl.0 .0);
+        rng.fill_bytes(&mut sr.0 .0);
+        // Use LSB of seed for ControlBits -- saves a call to the prng,
```

Comment:
Let's see if we can make this comment a little more useful. "call to prng" is a little vague, I think. What we're trying to do here concretely is get ensure that `prg()` costs exactly two calls to AES.

Also, the sentence needs to end with a period, not a comma.
```suggestion
        // Use LSB of seed for ControlBits -- saves a call to the prng,
```